### PR TITLE
DoGeneralOp convert vector to matrix

### PR DESCRIPTION
**Pull request by fortianyou**
This pull request is mainly to convert vector operands of DoGeneralOp into matrixes. Please focus on reviewing `dot_rewriter.cpp` since other TorchToMhlo conversions are modified just to make the unit tests reach the case of DotGeneral rewritten passes. All TorchToMhlo passes will be replaced by the upstream implementations from Torch-MLIR.

### CODE REVIEW

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 1]
```diff
@@ -1093,68 +1142,58 @@ class ConvertAtenMatmulBaseOp : public OpConversionPattern<AtenOpT> {
...
-
-    auto lhsShape = lhsTy.getShape();
-    auto rhsShape = rhsTy.getShape();
-
     auto lhsElemTy = lhsTy.getElementType();
     auto rhsElemTy = rhsTy.getElementType();
 
     if (lhsElemTy != rhsElemTy)
-      return op.emitError("Matmul: input datatypes mismatched");
+      return op.emitError("matmul: input datatypes mismatched");
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:1149]: nit: a bracket for `if` body?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 2]
```diff
@@ -1057,6 +1057,55 @@ LogicalResult ConvertAtenOp<AtenPowTensorScalarOp>::matchAndRewrite(
   return success();
 }
 
+void getBmmBroadcast(
+    PatternRewriter& rewriter,
+    Operation* op,
+    Value& inpLhs,
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:1063]: It's better to have output arguments at last ...

File: pytorch_blade/tests/disc/ops/test_matmul.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def _test_linear(self, dtype):
         self._test_disc(linear, [annotation, annotation], test_data)
         linear = torch.nn.Linear(256, 256).to(self.device).to(dtype)
         annotation = ([-1, 120, 256], dtype)
-        self._test_disc(linear, [annotation],  (x,), atol=1e-3, rtol=1e-3)
+        self._test_disc(linear, [annotation],  (x,), atol=5e-2, rtol=1e-3)
```
Comment (qiuxiafei) [inline pytorch_blade/tests/disc/ops/test_matmul.py:50]: Anything wrong? Is it `torch.half`?

File: pytorch_blade/tests/mhlo/matmul.mlir [hunk 4]
```diff
@@ -20,42 +21,44 @@ func.func @torch.aten.bmm(%arg0: !torch.vtensor<[10,3,4],f32>, %arg1: !torch.vte
...
 
 // -----
-// CHECK-LABEL:  func @torch.aten.bmm.dyn(
+
+// CHECK-LABEL:  func.func @torch.aten.bmm.dyn(
 // CHECK-SAME:         %[[ARG0:.*]]: tensor<4x?x256xf32>, %[[ARG1:.*]]: tensor<4x256x?xf32>) -> tensor<4x?x?xf32> {
-// CHECK:         %[[T0:.*]] = "mhlo.dot_general"(%[[ARG0]], %[[ARG1]]) {dot_dimension_numbers = #mhlo.dot<lhs_batching_dimensions = [0], rhs_batching_dimensions = [0], lhs_contracting_dimensions = [2], rhs_contracting_dimensions = [1]>} : (tensor<4x?x256xf32>, tensor<4x256x?xf32>) -> tensor<4x?x?xf32>
-// CHECK:         return %[[T0]] : tensor<4x?x?xf32>
+// CHECK:         %[[C256_I32:.*]] = arith.constant 256 : i32
+// CHECK:         %[[C4_I32:.*]] = arith.constant 4 : i32
```
Comment (qiuxiafei) [inline pytorch_blade/tests/mhlo/matmul.mlir:28]: maybe don't have to check the constants if not critical? I saw unstable order of in sequence of constant IR.

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 5]
```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
...
+  auto mhloShape = rewriter.create<tensor::FromElementsOp>(loc, newDimSizes);
+  return rewriter.create<mhlo::DynamicReshapeOp>(loc, outTy, tensor, mhloShape)
+      .getResult();
+}
+
+// Converts DotGeneralOp if it represents vec * vec, vec * mat, mat * vec.
+// Will first unsqueeze vector to matirx, then convert DoGeneralOp to
+// matrix multiply; After that drop(unsqueeze) the extra-dimensions from
+// the output of DoGeneralOp.
+struct DotGeneralConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:193]: Shall we have a .mlir test for this conversion on compiler side?

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 6]
```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
...
+  // Returns a new tensor with dims of size 1 inserted at the specified
+  // position.
+  //
+  // The position indices (must be high to low dimension number of the returned
+  // tensor) are specified with unsqzDims. Indices must be in-order, and in
+  // range of tensor rank. Thus, unsqueeze a rank 1 tensor with {0, 2}, {0, 1,
+  // 3}, {0, 1, 2} are all valid dimension sets, but {0, 3}, {2} are not.
+  auto dim_sizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto loc = op->getLoc();
+  auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:137]: nit: Are we fixing ConstantOp from arith and mhlo now?

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 7]
```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
...
+  auto rank = dim_sizes.size();
+
+  auto rankTy = tensor.getType().dyn_cast<RankedTensorType>();
+  auto oldShape = rankTy.getShape();
+
+  SmallVector<Value, 4> newDimSizes;
+  std::vector<int64_t> newShape;
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:144]: nit: just curious why `newShape` is not a `SmallVector` but `newDimSizes` is. :P

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 8]
```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
...
+  newDimSizes.reserve(rank - 1);
+  newShape.reserve(rank - 1);
+  for (size_t k = 0; k < rank; ++k) {
+    if (dim == k) continue;
+    newDimSizes.push_back(dim_sizes[k]);
+    newShape.push_back(oldShape[k]);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:179]: Should have a check that if the squeezed dim is of size `1`?

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc (hunk 5 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/dot_rewriter.cc:193]: +1

File: pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h [hunk 9]
```diff
@@ -218,6 +218,14 @@ mlir::Value getRollTensor(
     int64_t shiftInt,
     int64_t dim);
 
+Value getBroadcastTensor(
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h:221]: I note that the implementation of this function have a potential bug for rank-0 case.  we need to explicitly provide the output type when `dimSizes` is empty. Better to change the implementation as well. 
```
Value mhloShape = rewriter.create<tensor::FromElementsOp>(loc, dimSizes);
```

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 10]
```diff
@@ -1057,6 +1057,55 @@ LogicalResult ConvertAtenOp<AtenPowTensorScalarOp>::matchAndRewrite(
   return success();
 }
 
+void getBmmBroadcast(
```
[the remark this answers is not in the document]
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:1060]: bmm in PyTorch means BatchMatMul

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc [hunk 11]
```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
...
+    //
+    // unsqueeze lhs & rhs to matrix if needed
+    if (n_batch_dims == lhs_contracting_dims[0] && old_lhs_matrix_rank == 1) {
+      if (old_rhs_matrix_rank == 1) {
+        // vec x vec
+        new_lhs = unsqueezeTensorDim(rewriter, op, old_lhs, n_batch_dims);
+        new_rhs = unsqueezeTensorDim(rewriter, op, old_rhs, n_batch_dims + 1);
+        new_out_shape[n_batch_dims] = 1;
+        new_out_shape[n_batch_dims + 1] = 1;
+      } else if (n_batch_dims == lhs_contracting_dims[0]) {
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:248]: better to use a assert, instead of a `if` statement.

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc (hunk 6 above)
Comment (fortianyou) [reply tao_compiler/mlir/disc/transforms/dot_rewriter.cc:137]: Doesn't relate to MHLO, it requires a scalar constant here.

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc (hunk 11 above)
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/dot_rewriter.cc:248]: Do we need to check non-contracting/non-batch dimension of RHS is 1 as well? 
According to the doc, non-contracting/non-batch dimension may not always be one (e.g. from einsum)
https://www.tensorflow.org/xla/operation_semantics#dotgeneral

File: pytorch_blade/tests/disc/ops/test_matmul.py (hunk 3 above)
Comment (fortianyou) [reply pytorch_blade/tests/disc/ops/test_matmul.py]: fp16 precision error

File: pytorch_blade/tests/mhlo/matmul.mlir (hunk 4 above)
Comment (fortianyou) [reply pytorch_blade/tests/mhlo/matmul.mlir]: TorchMLIR checks all the lines of output IR.

File: tao_compiler/mlir/disc/transforms/dot_rewriter.cc (hunk 11 above)
Comment (fortianyou) [reply tao_compiler/mlir/disc/transforms/dot_rewriter.cc]: Only mhlo::Dot is put into consideration for now.

Comment (fortianyou) [reply tao_compiler/mlir/disc/transforms/dot_rewriter.cc]: Actually there is a bug before.

**Comment by fortianyou** [comment]
I removed the FileCheck test case because it triggered segfault on aarch64(run correctly on other platforms)
```
// -----

// CHECK-LABEL:  func.func @dot_1dx1d(
// CHECK-SAME:         %[[ARG0:.*]]: tensor<256xf32>, %[[ARG1:.*]]: tensor<256xf32>) -> tensor<f32> {
// CHECK:         %[[CST:.*]] = arith.constant dense<[1, 256]> : tensor<2xindex>
// CHECK:         %[[CST_0:.*]] = arith.constant dense<[256, 1]> : tensor<2xindex>
// CHECK:         %[[CST_1:.*]] = arith.constant dense<1> : tensor<1xindex>
// CHECK:         %[[CST_2:.*]] = arith.constant dense<> : tensor<0xindex>
// CHECK:         %[[T0:.*]] = "mhlo.dynamic_reshape"(%[[ARG0]], %[[CST]]) : (tensor<256xf32>, tensor<2xindex>) -> tensor<1x256xf32>
// CHECK:         %[[T1:.*]] = "mhlo.dynamic_reshape"(%[[ARG1]], %[[CST_0]]) : (tensor<256xf32>, tensor<2xindex>) -> tensor<256x1xf32>
// CHECK:         %[[T2:.*]] = "mhlo.dot_general"(%[[T0]], %[[T1]]) {dot_dimension_numbers = #mhlo.dot<lhs_contracting_dimensions = [1], rhs_contracting_dimensions = [0]>} : (tensor<1x256xf32>, tensor<256x1xf32>) -> tensor<1x1xf32>
// CHECK:         %[[T3:.*]] = "mhlo.dynamic_reshape"(%[[T2]], %[[CST_1]]) : (tensor<1x1xf32>, tensor<1xindex>) -> tensor<1xf32>
// CHECK:         %[[T4:.*]] = "mhlo.dynamic_reshape"(%[[T3]], %[[CST_2]]) : (tensor<1xf32>, tensor<0xindex>) -> tensor<f32>
// CHECK:         return %[[T4]] : tensor<f32>
func.func @dot_1dx1d(%arg0: tensor<256xf32>, %arg1: tensor<256xf32>) -> tensor<f32> {
  %0 = "mhlo.dot"(%arg0, %arg1) : (tensor<256xf32>, tensor<256xf32>) -> tensor<f32>
  return %0 : tensor<f32>
}
```

I think the segfault is not related to this PR. So I would like to dismiss this error to merge this PR quickly. And another bug report issue is created here https://github.com/alibaba/BladeDISC/issues/533

**Review by wyzero — APPROVED** [review]
LGTM
